impl/src/fmt.rs

```diff
@@ -94,6 +101,122 @@ impl Display<'_> {
         self.args = args;
         self.has_bonus_display = has_bonus_display;
     }
+
+    pub fn iter_fmt_types(&'a self, fields: &'a [Field]) -> Vec<DisplayFormatMarking<'a>> {
+        let members: Set<Member> = fields.iter().map(|f| f.member.clone()).collect();
+        let fmt = self.fmt.value();
+        let read = fmt.as_str();
+
+        let mut member_refs: Vec<(&Member, &Field, bool)> = Vec::new();
+
+        for template in parse_fmt_template(read) {
+            if let Some(target) = &template.target {
+                if members.contains(target) {
+                    if let Some(matching) = fields.iter().find(|f| &f.member == target) {
+                        member_refs.push((&matching.member, matching, template.is_display()));
+                    }
+                }
+            }
+        }
+
+        member_refs
+            .iter()
+            .map(|(_m, f, is_display)| {
+                if *is_display {
+                    DisplayFormatMarking::Display(*f)
+                } else {
+                    DisplayFormatMarking::Debug(*f)
+                }
+            })
+            .collect()
+    }
+}
+
+struct FormatInterpolation {
+    pub target: Option<Member>,
+    pub format: Option<String>,
+}
+
+impl FormatInterpolation {
+    pub fn is_debug(&self) -> bool {
+        self.format
+            .as_ref()
+            .map(|x| x.contains('?'))
+            .unwrap_or(false)
+    }
+
+    pub fn is_display(&self) -> bool {
+        !self.is_debug()
+    }
+}
+
+impl From<(Option<&str>, Option<&str>)> for FormatInterpolation {
+    fn from((target, style): (Option<&str>, Option<&str>)) -> Self {
+        let target = match target {
+            None => None,
+            Some(s) => Some(if let Ok(i) = s.parse::<usize>() {
+                Member::Unnamed(syn::Index {
+                    index: i as u32,
+                    span: Span::call_site(),
+                })
+            } else {
+                let mut s = s;
+                let ident = take_ident(&mut s);
+                Member::Named(ident)
+            }),
+        };
+        let format = style.map(String::from);
+        FormatInterpolation { target, format }
+    }
+}
+
+fn read_format_template(mut read: &str) -> Option<(FormatInterpolation, &str)> {
+    // If we aren't in a bracketed area, or we are in an escaped bracket, return None
+    if !read.starts_with('{') || read.starts_with("{{") {
+        return None;
+    }
+    // Read past the starting bracket
+    read = &read[1..];
+    // If there is no end bracket, bail
+    let end_bracket = read.find('}')?;
+    let contents = &read[..end_bracket];
+    let (name, style) = if let Some(colon) = contents.find(':') {
+        (&contents[..colon], &contents[colon + 1..])
+    } else {
+        (contents, "")
+    };
+
+    // Strip expanded identifier-prefixes since we just want the non-shorthand version
+    let name = if name.starts_with("field_") {
```

Comment:
This is a red flag &ndash; this code is trying to reverse engineer (lossily) information that already exists where the format string was already processed inside `expand_shorthand`.

Reply:
Agreed- that's a consequence of invoking it at the late-phase location I did- I think I probably should've augmented it by changing `expand_shorthand` to some sort of generic type that maps it out, but the problem was that my generic intersection code required knowledge of the generics before it could determine what _was_ a generic.

The majority of my time ended up sunk into writing parse methods using `syn::Parser`, only to find out that it couldn't handle imbalanced bracketing at any point, and didn't provide lower-level access than tokens. Having not spent much time with iterative parsers like the one in `expand_shorthand`, I couldn't see a straightforward way to extract the full information I needed from it without re-plumbing its state outputs.

Alas- as you showed- that was the right move ^^;